clients/go/admin/client.go

```diff
@@ -54,105 +78,153 @@ func GetAdditionalAdminClientConfigOptions(cfg Config) []grpc.DialOption {
-// It will first attempt to retrieve the token endpoint by making a metadata call. If that fails, but the token endpoint
-// is set in the config, that will be used instead.
-func getAuthenticationDialOption(ctx context.Context, cfg Config) (grpc.DialOption, error) {
-	var tokenURL string
-	tokenURL, err := getTokenEndpointFromAuthServer(ctx, cfg.AuthorizationServerURL)
-	if err != nil || tokenURL == "" {
-		logger.Infof(ctx, "No token URL found from configuration Issuer, looking for token endpoint directly")
-		if err != nil {
-			logger.Errorf(ctx, "Err is %s", err)
-		}
-		tokenURL = cfg.TokenURL
-		if tokenURL == "" {
-			return nil, errors.New("no token endpoint could be found")
-		}
+	clientMetadata, err := authClient.FlyteClient(ctx, &service.FlyteClientRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch client metadata. Error: %v", err)
 	}
 
 	secretBytes, err := ioutil.ReadFile(cfg.ClientSecretLocation)
 	if err != nil {
 		logger.Errorf(ctx, "Error reading secret from location %s", cfg.ClientSecretLocation)
 		return nil, err
 	}
+
 	secret := strings.TrimSpace(string(secretBytes))
 
+	scopes := cfg.Scopes
+	if len(scopes) == 0 {
+		scopes = clientMetadata.Scopes
+	}
+
 	ccConfig := clientcredentials.Config{
 		ClientID:     cfg.ClientID,
 		ClientSecret: secret,
 		TokenURL:     tokenURL,
-		Scopes:       cfg.Scopes,
+		Scopes:       scopes,
 	}
+
 	tSource := ccConfig.TokenSource(ctx)
-	oauthTokenSource := NewCustomHeaderTokenSource(tSource, cfg.AuthorizationHeader)
+	oauthTokenSource := NewCustomHeaderTokenSource(tSource, clientMetadata.AuthorizationMetadataKey)
 	return grpc.WithPerRPCCredentials(oauthTokenSource), nil
 }
 
-func NewAdminConnection(ctx context.Context, cfg Config) (*grpc.ClientConn, error) {
-	var opts []grpc.DialOption
+// InitializeAuthMetadataClient creates a new anonymously Auth Metadata Service client.
+func InitializeAuthMetadataClient(ctx context.Context, cfg *Config) (client service.AuthMetadataServiceClient, err error) {
+	onceAuthMetadata.Do(func() {
+		authMetadataConnection, err = NewAdminConnection(ctx, cfg)
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize admin connection. Error: %w", err)
+	}
+
+	return service.NewAuthMetadataServiceClient(authMetadataConnection), nil
+}
+
+func NewServiceAuthDialOptions(ctx context.Context, cfg *Config, authClient service.AuthMetadataServiceClient) ([]grpc.DialOption, error) {
+	if cfg.UseAuth {
+		logger.Infof(ctx, "Instantiating a token source to authenticate against Admin, ID: %s", cfg.ClientID)
+		opt, err := getAuthenticationDialOption(ctx, cfg, authClient)
+		if err != nil {
+			return nil, err
+		}
+
+		return []grpc.DialOption{opt}, nil
+	}
+
+	return nil, nil
+}
+
+func NewAdminConnection(_ context.Context, cfg *Config, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	if opts == nil {
+		opts = make([]grpc.DialOption, 0, 10)
```

Comment:
nit : May be somewhere we can mention the reasoning for this size which can act as a reference for required set of opts for authorization versus normal ones when talking to admin

Reply:
done